ledger/eval/eval.go

```diff
@@ -1547,8 +1581,35 @@ func isAbsent(totalOnlineStake basics.MicroAlgos, acctStake basics.MicroAlgos, l
 		return false
 	}
 	// See if the account has exceeded 10x their expected observation interval.
-	allowableLag := basics.Round(10 * totalOnlineStake.Raw / acctStake.Raw)
-	return lastSeen+allowableLag < current
+	allowableLag, o := basics.Muldiv(10, totalOnlineStake.Raw, acctStake.Raw)
+	if o {
+		// This can't happen with 10B total possible stake, but if we imagine
+		// another algorand network with huge possible stake, this seems reasonable.
+		allowableLag = math.MaxInt64 / acctStake.Raw
+	}
+	return lastSeen+basics.Round(allowableLag) < current
+}
+
+func (eval *BlockEvaluator) activeChallenge() challenge {
+	current := eval.Round()
+	var round basics.Round
+	if current > challengeInterval {
+		lastChallenge := current - (current % challengeInterval)
+		// challengeRound is in effect if we're after one grace period, but before the 2nd ends.
+		if current > lastChallenge+challengeGracePeriod && current < lastChallenge+2*challengeGracePeriod {
+			round = lastChallenge
+			challengeHdr, err := eval.state.BlockHdr(round)
+			if err != nil {
+				panic(err)
+			}
+			return challenge{round, challengeHdr.Seed, challengeBits}
+		}
+	}
+	return challenge{}
+}
+
+func failsChallenge(ch challenge, address basics.Address, lastSeen basics.Round) bool {
+	return ch.round != 0 && bitsMatch(ch.seed[:], address[:], ch.bits) && lastSeen < ch.round
```

Comment:
I think it's important enough to fix, because you can be marked absent even if you are not eligible for incentives.  It's true that every account ought to know how to get back on quickly after the heartbeat PR, it still seems weird someone could get knocked off immediately after upgrade.

I will make it so challenges don't count if they happened with different mining rules in effect (that is, an upgrade happened that changed the rules)

Reply:
Fixed in last commit. Unfortunately the code got a lot uglier, poking around like this, and making it unit testable. But such is life.